test/hw4/test_task2.py

```diff
@@ -0,0 +1,38 @@
+from unittest.mock import patch
+
+import pytest
+
+from All_home_works.hw4.task2 import count_dots_on_i
+
+URL = "https://example.com/"
+
+
+class Response:
+    def __init__(self, data: str):
+        self.data = data.split(" ")
+
+    def __enter__(self):
+        return map(lambda x: x.encode(), self.data)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        return None
+
+
+@patch("urllib.request.urlopen")
+def test_urlopen__with_right_url(urlopen):
```

Comment:
You can join it with `test_count_dots_on_i__check_valid_data`

- `@patch(` is the preparation step (`Given`)
- `count_dots_on_i` is our action (`When`)
- mock call and return value are the result of the action that should be asserted (`Then`). 

In that case your action is complex and this is ok to test all it's aspects in the single test.